taichi/runtime/opengl/opengl_api.cpp

```diff
@@ -400,23 +401,26 @@ void DeviceCompiledTaichiKernel::launch(RuntimeContext &ctx,
   for (auto &item : program_.arr_args) {
     int i = item.first;
     TI_ASSERT(args[i].is_array);
-    if (args[i].size == 0 || ctx.is_device_allocation[i])
+    const auto &arr_meta = ctx.array_metadata[i];
+    const auto arr_runtime_sz = arr_meta.runtime_size;
+    if ((arr_runtime_sz == 0) || arr_meta.is_device_allocation) {
       continue;
+    }
     has_ext_arr = true;
-    if (args[i].size != item.second.total_size ||
+    if (arr_runtime_sz != item.second.runtime_size ||
```

Comment:
I think we should do this only when `arr_runtime_sz > item.second.runtime_size`? @ailzhang @bobcao3

Reply:
Yup I think that'll work! Note this behavior differs by backend as Vulkan always deallocate the memory allocated for external arrays.  https://github.com/taichi-dev/taichi/blob/master/taichi/runtime/vulkan/runtime.cpp#L547-L554 We can probably align this behavior across them as well.

Reply:
+1